File: src/tutorial/src/data_logger.cpp
#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <sensor_msgs/NavSatFix.h>
#include <geometry_msgs/TwistWithCovarianceStamped.h>
#include <tf/transform_datatypes.h>
#include <fstream>
#include <iomanip>
#include <string>
#include <vector>
#include <sstream>
#include <mutex>
#include <thread>
#include <ctime>

class DataLogger
{
private:
    ros::NodeHandle nh_;
    ros::Subscriber gss_sub_;
    ros::Subscriber gps_sub_;
    ros::Subscriber imu_sub_;

    std::mutex buffer_mutex_;
    std::vector<std::vector<std::string>> buffer_;

    double gss_linear_x_, gss_linear_y_, gss_linear_z_;
    double gss_angular_x_, gss_angular_y_, gss_angular_z_;
    double lat_, lon_, alt_;
    double ax_, ay_, az_;
    double roll_, pitch_, yaw_;
    double r_rate_, p_rate_, y_rate_;

    double initial_time_;
    std::string log_file_;
    ros::Rate rate_;

public:
    DataLogger() : rate_(100)
    {
        gss_sub_ = nh_.subscribe("/fsds/gss", 10, &DataLogger::gssCallback, this);
        gps_sub_ = nh_.subscribe("/fsds/gps", 10, &DataLogger::gpsCallback, this);
        imu_sub_ = nh_.subscribe("/fsds/imu", 10, &DataLogger::imuCallback, this);

        initial_time_ = ros::Time::now().toSec();

        std::time_t t = std::time(nullptr);
        std::tm *now = std::localtime(&t);
        std::ostringstream filename;
        filename << "/home/user/fsds_ws/src/tutorial/log/data_log_" 
                 << std::put_time(now, "%m%d%H%M") << ".csv";
        log_file_ = filename.str();

        std::ofstream file(log_file_);
        file << "time[sec],ax[m/s^2],ay[m/s^2],az[m/s^2],roll_rate[deg/s],pitch_rate[deg/s],yaw_rate[deg/s],"
             << "roll[deg],pitch[deg],yaw[deg],lat[deg],lon[deg],alt[m],"
             << "gss_linear_x[m/s],gss_linear_y[m/s],gss_linear_z[m/s],"
             << "gss_angular_x[rad/s],gss_angular_y[rad/s],gss_angular_z[rad/s]\n";
        file.close();

        std::thread(&DataLogger::writeDataLog, this).detach();
        ROS_INFO("DATA LOGGING");
    }

    void gssCallback(const geometry_msgs::TwistWithCovarianceStamped::ConstPtr &msg)
    {
        gss_linear_x_ = msg->twist.twist.linear.x;
        gss_linear_y_ = msg->twist.twist.linear.y;
        gss_linear_z_ = msg->twist.twist.linear.z;

        gss_angular_x_ = msg->twist.twist.angular.x;
        gss_angular_y_ = msg->twist.twist.angular.y;
        gss_angular_z_ = msg->twist.twist.angular.z;
    }

    void gpsCallback(const sensor_msgs::NavSatFix::ConstPtr &msg)
    {
        lat_ = msg->latitude;
        lon_ = msg->longitude;
        alt_ = msg->altitude;
    }

    void imuCallback(const sensor_msgs::Imu::ConstPtr &msg)
    {
        ax_ = msg->linear_acceleration.x;
        ay_ = msg->linear_acceleration.y;
        az_ = msg->linear_acceleration.z;

        tf::Quaternion q(
            msg->orientation.x,
            msg->orientation.y,
            msg->orientation.z,
            msg->orientation.w);
        tf::Matrix3x3 m(q);
        double roll, pitch, yaw;
        m.getRPY(roll, pitch, yaw);

        roll_ = roll * 180.0 / M_PI;
        pitch_ = pitch * 180.0 / M_PI;
        yaw_ = yaw * 180.0 / M_PI;

        r_rate_ = msg->angular_velocity.x;
        p_rate_ = msg->angular_velocity.y;
        y_rate_ = msg->angular_velocity.z;
    }

    void writeDataLog()
    {
        ros::Rate loop_rate(10); // Write every 0.1s
        while (ros::ok())
        {
            logData();
            loop_rate.sleep();
        }
    }

    void logData()
    {
        double t = ros::Time::now().toSec() - initial_time_;
        std::vector<std::string> data_row;
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(2) << t;
        data_row.push_back(ss.str());

        auto add_data = [&](double value, int precision = 5) {
            std::ostringstream s;
            s << std::fixed << std::setprecision(precision) << value;
            data_row.push_back(s.str());
        };

        add_data(ax_);
        add_data(ay_);
        add_data(az_);
        add_data(r_rate_);
        add_data(p_rate_);
        add_data(y_rate_);
        add_data(roll_);
        add_data(pitch_);
        add_data(yaw_);
        add_data(lat_, 8);
        add_data(lon_, 8);
        add_data(alt_, 3);
        add_data(gss_linear_x_);
        add_data(gss_linear_y_);
        add_data(gss_linear_z_);
        add_data(gss_angular_x_);
        add_data(gss_angular_y_);
        add_data(gss_angular_z_);

        {
            std::lock_guard<std::mutex> lock(buffer_mutex_);
            buffer_.push_back(data_row);
        }

        flushBuffer();
    }

    void flushBuffer()
    {
        std::lock_guard<std::mutex> lock(buffer_mutex_);
        if (!buffer_.empty())
        {
            std::ofstream file(log_file_, std::ios::app);
            for (const auto &row : buffer_)
            {
                for (size_t i = 0; i < row.size(); ++i)
                {
                    file << row[i];
                    if (i != row.size() - 1)
                        file << ",";
                }
                file << "\n";
            }
            buffer_.clear();
        }
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "Data_Logger");
    DataLogger logger;
    ros::spin();
    return 0;
}

File: src/tutorial/scripts/data_logger.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

#####################
# 수집 대상 데이터 : 
# time, ax, ay, az, roll, pitch, yaw,
# x, y, z,
# motor rpm, wheel angle
#
#####################
import rospy, rospkg
import tf
import os
import threading
import csv
from std_msgs.msg import Float32MultiArray, Float64
from sensor_msgs.msg import Imu, NavSatFix
from ublox_msgs.msg import NavSAT
from geometry_msgs.msg import TwistWithCovarianceStamped
from nav_msgs.msg import Odometry
from tf.transformations import euler_from_quaternion
from pyproj import Proj
from math import pi, sqrt
from datetime import datetime
import numpy as np
import time
class DataLogger:
    def __init__(self):
        rospy.init_node("Data_Logger", anonymous=True)
        self.gss_sub = rospy.Subscriber("/fsds/gss",TwistWithCovarianceStamped, self.gss_callback)
        self.gps_sub = rospy.Subscriber("/fsds/gps",NavSatFix, self.gps_callback)
        self.imu_sub = rospy.Subscriber("/fsds/imu",Imu, self.imu_callback)

        ## gss
        self.gss_linear_x , self.gss_linear_y, self.gss_linear_z = 0,0,0
        self.gss_angular_x , self.gss_angular_y, self.gss_angular_z = 0,0,0

        ## gps
        self.lat, self.lon, self.alt = 0,0,0

        ## imu
        self.ax, self.ay, self.az = 0,0,0
        self.roll, self.pitch, self.yaw = 0,0,0
        self.r_rate, self.p_rate, self.y_rate = 0,0,0

        self.initial_time = rospy.get_time()

        self.today = datetime.today().strftime("%m%d%H%M")
        rp = rospkg.RosPack()
        package_path = rp.get_path('tutorial')
        log_dir = os.path.join(package_path,"log")
        os.makedirs(log_dir, exist_ok=True)
        self.buffer = []
        self.buffer_lock = threading.Lock()
        
        self.data_log_file = os.path.join(log_dir,f"data_log_{self.today}.csv")
        self.rate = rospy.Rate(100)
        with open(self.data_log_file, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["time[sec]","ax[m/s^2]","ay[m/s^2]","az[m/s^2]","roll_rate[deg/s]","pitch_rate[deg/s]","yaw_rate[deg/s]","roll[deg]","pitch[deg]","yaw[deg]","lat[deg]","lon[deg]","alt[m]","gss_linear_x[m/s]","gss_linear_y[m/s]","gss_linear_z[m/s]","gss_angular_x[rad/s]","gss_angular_y[rad/s]","gss_angular_z[rad/s]"])
        
        threading.Thread(target=self.write_data_log, daemon=True).start()
        while not rospy.is_shutdown():
            # print("Logging in progress ...")
            self.write_data_log()
            # rate.sleep()
            # self.data_logger()
            # print(self.cov_mat)
        rospy.spin()
        
    def gss_callback(self,data):
        gss_linear = data.twist.twist.linear
        gss_angular = data.twist.twist.angular

        self.gss_linear_x = gss_linear.x
        self.gss_linear_y = gss_linear.y
        self.gss_linear_z = gss_linear.z

        self.gss_angular_x = gss_angular.x
        self.gss_angular_y = gss_angular.y
        self.gss_angular_z = gss_angular.z


    def gps_callback(self,data):
        self.lat = data.latitude
        self.lon = data.longitude
        self.alt = data.altitude

    def imu_callback(self,data):
        ## Linear Acceleration
        self.ax = data.linear_acceleration.x
        self.ay = data.linear_acceleration.y
        self.az = data.linear_acceleration.z
        
        ## angle
        q = data.orientation
        quaternion = (q.x, q.y, q.z, q.w)
        r, p, y = euler_from_quaternion(quaternion)
        self.roll = r * 180.0 / pi
        self.pitch = p * 180.0 / pi
        self.yaw = y * 180.0/pi

        ## angular velocity
        self.r_rate = data.angular_velocity.x
        self.p_rate = data.angular_velocity.y
        self.y_rate = data.angular_velocity.z

    def data_logger(self):
        # t_old = 0
        t = rospy.get_time() - self.initial_time
        # print(t)
        # self.enc_log_data = np.append(self.enc_log_data,[[t,real_steer,desire_steer,pwm,vel,state,current_vel,con_sp,desired_vel]],axis=0)
        self.buffer.append([round(t,1),self.ax,self.ay,self.az,self.roll,self.pitch,self.yaw,self.r_rate,self.p_rate,self.y_rate,self.lat,self.lon,self.alt,self.gss_linear_x,self.gss_linear_y,self.gss_linear_z,self.gss_angular_x,self.gss_angular_y,self.gss_angular_z])
        self.rate.sleep()
        print(self.buffer)
    def write_data_log(self):
        while not rospy.is_shutdown():
            self.data_logger()
            print("LOGGING")
            with self.buffer_lock:
                buffer_copy = self.buffer[:]
                self.buffer = []
                if buffer_copy:
                    with open(self.data_log_file, 'a', newline='') as f:
                        writer = csv.writer(f)
                        writer.writerows(buffer_copy)
if __name__ == '__main__':
    try:
        
        EncoderLog = DataLogger()
    except rospy.ROSInterruptException:
        pass